rzup/install

```diff
@@ -1,54 +1,214 @@
 
-command -v curl >/dev/null 2>&1 || { echo >&2 "curl is required but it's not installed. Aborting."; exit 1; }
+usage() {
+  cat <<EOF
+install
 
-# Create the .rzup bin directory and rzup binary if it doesn't exist.
-mkdir -p $RISC0_BIN_DIR
-curl -# -L $BIN_URL -o $BIN_PATH
-chmod +x $BIN_PATH
+The installer for rzup, from RISC Zero.
 
-detect_shell() {
-    case $SHELL in
-    */zsh) PROFILE="${ZDOTDIR:-"$HOME"}/.zshenv"; PREF_SHELL='zsh' ;;
-    */bash) PROFILE="$HOME/.bashrc"; PREF_SHELL='bash' ;;
-    */fish) PROFILE="$HOME/.config/fish/config.fish"; PREF_SHELL='fish' ;;
-    */ash) PROFILE="$HOME/.profile"; PREF_SHELL='ash' ;;
-    *) echo "Could not detect shell, manually add ${RISC0_BIN_DIR} to your PATH." && exit 1 ;;
+Usage: rzup-init.sh [OPTIONS]
+
+Options:
+  -v, --verbose
+          Enable verbose output
+  -q, --quiet
+          Disable progress output
+  -h, --help
+          Print help
+EOF
+}
+
+cleanup() {
+  [ -n "${_dir:-}" ] && [ -d "$_dir" ] && rm -rf "$_dir"
+}
+
+trap cleanup EXIT
+
+main() {
+  downloader --check
+  need_cmd uname
+  need_cmd mktemp
+  need_cmd chmod
+  need_cmd mkdir
+  need_cmd rm
+  need_cmd rmdir
+  need_cmd mv
+
+  check_rust_installed
+  get_architecture
+
+  _arch="$RETVAL"
+  assert_nz "$_arch" "arch"
+
+  _url="${RZUP_URL}${TEST_PATH}${_arch}/rzup"
+  _dir="$(ensure mktemp -d)"
+  _file="${_dir}/rzup"
+
+  [ -f "$RZUP_BIN_DIR/rzup" ] && rm "$RZUP_BIN_DIR/rzup"
+
+  for arg in "$@"; do
+    case "$arg" in
+    --help)
+      usage
+      exit 0
+      ;;
+    -q | --quiet) QUIET=yes ;;
+    *) ;;
     esac
+  done
+
+  info 'Downloading installer'
+
+  ensure mkdir -p "$_dir"
+  ensure downloader "$_url" "$_file"
+  ensure chmod u+x "$_file"
+  [ ! -x "$_file" ] && err "Cannot execute $_file"
+
+  ensure mkdir -p "$RZUP_BIN_DIR"
+  ensure mv "$_file" "$RZUP_BIN_DIR"
+
+  info "rzup has been installed to $RZUP_BIN_DIR"
+
+  detect_shell
+
+  # add rzup to PATH if it isn't already present
+  if [[ ":$PATH:" != *":${RZUP_BIN_DIR}:"* ]]; then
+      info "Adding rzup to PATH in ${PROFILE}"
+      {
+          echo
+          echo "export PATH=\"\$PATH:$RZUP_BIN_DIR\""
+      } >> "$PROFILE"
+  else
+      info "rzup found in PATH"
+  fi
+
+  info "🎉 rzup installed!"
+  echo "Run the following command to install the zkVM:"
+  echo "  rzup install"
+}
+
+check_rust_installed() {
+  check_cmd rustc || err "Rust is not installed. Please install Rust from https://rustup.rs/ and run this script again."
+}
+
+get_architecture() {
+  _ostype="$(uname -s)"
+  _cputype="$(uname -m)"
+
+  case "$_ostype" in
+  Linux) _ostype=linux ;;
+  Darwin) _ostype=darwin ;;
+  *) err "Unsupported OS type: $_ostype" ;;
+  esac
+
+  case "$_cputype" in
+  x86_64 | amd64) _cputype=x86_64 ;;
+  aarch64 | arm64) _cputype=arm64 ;;
+  *) err "Unsupported CPU type: $_cputype" ;;
+  esac
+
+  case "$_cputype $_ostype" in
+  "x86_64 linux") _arch=Linux-X64-cpu ;;
+  "arm64 darwin" ) _arch=macOS-ARM64-cpu ;;
```

Comment:
```suggestion
  "x86_64 linux") _arch=Linux-X64 ;;
  "arm64 darwin" ) _arch=macOS-ARM64 ;;
```

Reply:
I'll change the s3 bucket to reflect this